haskell/test/Test/Invariants.hs

```diff
@@ -8,22 +8,90 @@ module Test.Invariants
 import Prelude
 import qualified Unsafe
 
+import Data.Ix (inRange)
 import qualified Data.List as List
 import qualified Data.Map as Map
 import Fmt
 import Lorentz hiding (assert, not, now, (>>))
 import Morley.Nettest
+import Tezos.Core (timestampToSeconds)
 
 import SegCFMM.Types
 import Test.Math
 import Test.Util
 
-checkAllInvariants :: (HasCallStack, MonadEmulated caps base m) => ContractHandler cp Storage -> m ()
-checkAllInvariants ch = do
-  st <- getFullStorage ch
+checkAllInvariants :: (HasCallStack, MonadEmulated caps base m) => ContractHandler Parameter Storage -> m ()
+checkAllInvariants cfmm = do
+  st <- getFullStorage cfmm
   checkTickMapInvariants st
   checkTickInvariants st
   checkStorageInvariants st
+  checkBalanceInvariants cfmm st
+  checkAccumulatorsInvariants cfmm st
+
+-- | Invariants:
+-- 1. For all accumulators: the sum of all the values accumulated between any two consecutive ticks
+--    must equal the global accumulated value.
+--    E.g.: if the ticks [i1, i2, i3, i4] have been initialized, then:
+--      fee_growth == fee_growth_inside(i1, i2) + fee_growth_inside(i2, i3) + fee_growth_inside(i3, i4)
+{-# ANN checkAccumulatorsInvariants ("HLint: ignore Use uncurry" :: Text) #-}
+checkAccumulatorsInvariants
+  :: (HasCallStack, MonadEmulated caps base m)
+  => ContractHandler Parameter Storage
+  -> Storage
+  -> m ()
+checkAccumulatorsInvariants cfmm st = do
+  tickIndices <- mapToList (sTicks st) <&> fmap fst
+  let tickIndicesPaired = tickIndices `zip` Unsafe.tail tickIndices
+  insideAccumulators <- for tickIndicesPaired \(idx1, idx2) -> tickAccumulatorsInside cfmm st idx1 idx2
+  let sumInsideAccumulators = foldl1 addTickAccumulators insideAccumulators
+
+  CumulativesValue {cvTickCumulative, cvSecondsPerLiquidityCumulative} <- observe cfmm
+  now <- getNow
+  let globalAccumulators =
+        Accumulators
+          { aSeconds = timestampToSeconds now
+          , aTickCumulative = cvTickCumulative
+          , aFeeGrowth = sFeeGrowth st
+          , aSecondsPerLiquidity = cvSecondsPerLiquidityCumulative
+          }
+
+  globalAccumulators @== sumInsideAccumulators
+  where
+    addTickAccumulators :: Accumulators -> Accumulators -> Accumulators
+    addTickAccumulators tc1 tc2 =
+      Accumulators
+        { aSeconds = aSeconds tc1 + aSeconds tc2
+        , aTickCumulative = aTickCumulative tc1 + aTickCumulative tc2
+        , aFeeGrowth = aFeeGrowth tc1 + aFeeGrowth tc2
+        , aSecondsPerLiquidity = aSecondsPerLiquidity tc1 + aSecondsPerLiquidity tc2
+        }
+
+-- | Invariants:
+-- 1. The contract always has enough balance to liquidite all positions (and pay any fees due).
+checkBalanceInvariants :: (HasCallStack, MonadEmulated caps base m) => ContractHandler Parameter Storage -> Storage -> m ()
```

Comment:
Minor nitpick, but I suppose that this and the one in `checkAccumulatorsInvariants` don't need to be "listed".

Unless you want to count them all in their order in `checkAllInvariants`, but in that case the numbers should be changed.

Reply:
Done